packages/repository/src/model.ts

```diff
@@ -263,6 +263,31 @@ export abstract class Entity extends Model implements Persistable {
   }
 }
 
+/**
+ * An anonymous function that resolves to a class/entity
+ * Intended to be used for cases when the JS engine is unable to fully define
+ * a given type
+ */
+export type TypeResolver<T = typeof Entity> = () => T;
+
+/**
+ * A function that checks whether given element is a TypeResolver or not
+ * @param fn
+ */
+export function isTypeResolver<T>(
+  fn: TypeResolver<T> | T,
+): fn is TypeResolver<T> {
+  return !/^class/.test(fn.toString());
```

Comment:
Cross-posting https://github.com/strongloop/loopback-next/pull/1618#discussion_r212274216.

Hmm, this is not very reliable, is it?

Once we add support for JavaScript, people will be able to write models using the pre-ES6 way where no `class` keyword is used.

```ts
function MyModel(data) {
  Entity.call(this, data);
  // the rest of the constructor
}

util.inherits(MyModel, Entity);
```

If a model/entity class has to always inherit from our Model/Entity, then it may be more robust to check `fn.prototype` inheritance chain, something along the lines

```ts
return !(fn.prototype instanceof Model)
```

This assumes the `Model` constructor was loaded by the application from the same instance of `@loopback/repository` as our `isModelResolver` function was loaded from, which may not be always the case.

A more robust option is to crawl inheritance chain and check if a constructor name matches `Model`.

```ts
for (proto = fn.prototype; proto != null, proto = Object.getPrototypeOf(proto)) {
  if (proto.constructor.name === 'Model') return true;
}
return false;
```

To me, this complexity is a signal that we need to look for a different API design, one where we don't need to distinguish between a model constructor function and a model provider function.

---

❗️ By making this type more generic, i.e. a provider for any type, it becomes even more difficult to distinguish between a type and a provider. For example, what if the type is a function? Or a built-in type like `Date`, `RegExp` or `Buffer`?

```
$ node
> RegExp.toString()
'function RegExp() { [native code] }'
> Date.toString()
'function Date() { [native code] }'
> Buffer.toString()
'function Buffer(arg, encodingOrOffset, length) {\n  showFlaggedDeprecation();\n  // Common case.\n  if (typeof arg === \'number\') {\n    if (typeof encodingOrOffset === \'string\') {\n      throw new ERR_INVALID_ARG_TYPE(\'string\', \'string\', arg);\n    }\n    return Buffer.alloc(arg);\n  }\n  return Buffer.from(arg, encodingOrOffset, length);\n}'
```

Reply:
I'm personally fine with the complex proposal you've given here, mainly because I can't think of a better alternative API design that wouldn't need to differentiate between a constructor and a function returning a constructor. In regards to addressing built-in types like `Date`, I think that can be hard coded to check for, and addressing plain functions might be doable with another complex and hacky check.

Do you have an alternative in mind? I've been mulling over a way to support the relation decorators taking in either the resolver or the model constructor, but I haven't come up with a clean alternative.

Reply:
My personal conclusion is that we need to modify the design of relation decorators so that we don't need to distinguish between a type (a model class) and a type provider (a model class provider). This is one of the reason why our `Binding` class has so many different methods for setting the value (`to()`, `toDynamicValue()`, `toClass` and `toProvider`) - JavaScript is not powerful enough to allow code to differentiate between these different types at runtime.